SmartDeviceLink/SDLChoiceSet.m

```diff
@@ -0,0 +1,129 @@
+}
+
+- (instancetype)initWithTitle:(NSString *)title delegate:(id<SDLChoiceSetDelegate>)delegate layout:(SDLChoiceSetLayout)layout timeout:(NSTimeInterval)timeout initialPromptString:(nullable NSString *)initialPrompt timeoutPromptString:(nullable NSString *)timeoutPrompt helpPromptString:(nullable NSString *)helpPrompt vrHelpList:(nullable NSArray<SDLVRHelpItem *> *)helpList choices:(NSArray<SDLChoiceCell *> *)choices {
+    NSArray<SDLTTSChunk *> *initialTTS = [SDLTTSChunk textChunksFromString:initialPrompt];
+    NSArray<SDLTTSChunk *> *timeoutTTS = [SDLTTSChunk textChunksFromString:timeoutPrompt];
+    NSArray<SDLTTSChunk *> *helpTTS = [SDLTTSChunk textChunksFromString:helpPrompt];
+
+    return [self initWithTitle:title delegate:delegate layout:layout timeout:timeout initialPrompt:initialTTS timeoutPrompt:timeoutTTS helpPrompt:helpTTS vrHelpList:helpList choices:choices];
+}
+
+- (instancetype)initWithTitle:(NSString *)title delegate:(id<SDLChoiceSetDelegate>)delegate layout:(SDLChoiceSetLayout)layout timeout:(NSTimeInterval)timeout initialPrompt:(nullable NSArray<SDLTTSChunk *> *)initialPrompt timeoutPrompt:(nullable NSArray<SDLTTSChunk *> *)timeoutPrompt helpPrompt:(nullable NSArray<SDLTTSChunk *> *)helpPrompt vrHelpList:(nullable NSArray<SDLVRHelpItem *> *)helpList choices:(NSArray<SDLChoiceCell *> *)choices {
+    self = [self init];
+    if (!self) { return nil; }
+
+    if (choices.count == 0 || choices.count > 100) {
+        SDLLogW(@"Attempted to create a choice set with %lu choices; Only 1 - 100 choices are valid", choices.count);
+        return nil;
+    }
+
+    if (timeout < 5 || timeout > 100) {
+        SDLLogW(@"Attempted to create a choice set with a %f second timeout; Only 5 - 100 seconds is valid", timeout);
+        return nil;
+    }
+
+    if (title.length == 0 || title.length > 500) {
+        SDLLogW(@"Attempted to create a choice set with a %lu length. Only 500 characters are supported", title.length);
+        return nil;
+    }
+
+    NSMutableSet<NSString *> *choiceTextSet = [NSMutableSet setWithCapacity:choices.count];
+    for (SDLChoiceCell *cell in choices) {
+        [choiceTextSet addObject:cell.text];
+    }
+    if (choiceTextSet.count < choices.count) {
+        SDLLogW(@"Attempted to create a choice set with duplicate cell text. Cell text must be unique.");
+        return nil;
+    }
+
+    NSMutableArray<NSArray<NSString *> *> *nonNilVoiceCommands = [NSMutableArray array];
+    for (SDLChoiceCell *cell in choices) {
+        if (cell.voiceCommands != nil) {
+            [nonNilVoiceCommands addObject:cell.voiceCommands];
+        }
+    }
+
+    NSMutableSet<NSArray<NSString *> *> *choiceVoiceCommandSet = [NSMutableSet setWithArray:nonNilVoiceCommands];
+    if (choiceVoiceCommandSet.count < nonNilVoiceCommands.count) {
+        SDLLogW(@"Attempted to create a choice set with duplicate voice commands. Voice commands must be unique.");
+        return nil;
+    }
```

Comment:
I think it would be better if the data validation was done in the `SDLChoiceSetManager`'s `preloadChoices:withCompletionHandler` method. Currently:
1. If the `SDLChoiceCell` is `null`, the only way to get the error is to look at the logs.
2. It is still possible to send invalid data if setting each of params individually.

If data validation is done in the `preloadChoices:withCompletionHandler` method, the custom error can be passed back via the handler.

Reply:
We shouldn't do it in `preloadChoices:withCompletionHandler:` because we can upload multiple choices with the same primary text, but we just can't display a choice set with 2 that have the same text. I believe it's the same for voice commands.

Reply:
True, it is the most efficient way to validate the choice sets since the manager has 3 choice set upload methods.